### PR TITLE
[JBPM-9675] Find pot-owners as OR

### PR DESCRIPTION
**Pull request by ruromero**
Signed-off-by: ruromero <rromerom@redhat.com>

**JIRA**: JBPM-9675

[https://issues.redhat.com/browse/JBPM-9675](https://issues.redhat.com/browse/JBPM-9675)

### CODE REVIEW

File: jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/AbstractAdvanceRuntimeDataServiceImpl.java [hunk 1]
```diff
@@ -196,7 +196,6 @@ public void setEmf(EntityManagerFactory emf) {
                              "           FROM PeopleAssignments_PotOwners po \n" +
                              "           WHERE po.entity_id IN (:owners) \n" +
                              "           GROUP BY po.task_id \n" +
-                             "           HAVING COUNT(po.entity_id) = :num_owners \n" +
```
Comment (elguardian) [inline jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/AbstractAdvanceRuntimeDataServiceImpl.java:199]: Hi this is breaking AND operator and more likely to break backward compatibility.

Comment (elguardian) [reply jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/AbstractAdvanceRuntimeDataServiceImpl.java:199]: you need to add some sort of information to provide to the function whether it should be execute and AND or and OR

Comment (ruromero) [reply jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/AbstractAdvanceRuntimeDataServiceImpl.java:199]: How do you suggest to add this AND/OR option without modifying the existing API? Maybe create a new method?

Comment (fjtirado) [reply jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/AbstractAdvanceRuntimeDataServiceImpl.java:199]: @ruromero  I think you can check for the presence  a "new" QueryParam in attributes args called, for example "CONDITION", which can take values "AND" or "OR", as done here https://github.com/kiegroup/jbpm/blob/master/jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/AbstractAdvanceRuntimeDataServiceImpl.java#L115

Comment (ruromero) [reply jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/AbstractAdvanceRuntimeDataServiceImpl.java:199]: Thanks for the suggestion @fjtirado I have added the queryParam but I have reused the MODE. Does it make sense?

Comment (elguardian) [reply jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/AbstractAdvanceRuntimeDataServiceImpl.java:199]: the right approach would be to transform this one 
 List<String> ownersArg
to 
QueryParam ownersArg
two new operandas for collections:
ANY (in the collection)
ALL (all in the collection)

reimplement current methods as QueryParam( list of owners)
ANY = OR
ALL = AND

Comment (elguardian) [reply jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/AbstractAdvanceRuntimeDataServiceImpl.java:199]: and add two new methods accepting the query param

Comment (ruromero) [reply jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/AbstractAdvanceRuntimeDataServiceImpl.java:199]: @elguardian let me know if this is what you had in mind and I'll proceed with the droolsjbpm-integration side

File: jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/AdvanceRuntimeDataServiceImpl.java [hunk 2]
```diff
@@ -53,7 +53,7 @@ public AdvanceRuntimeDataServiceImpl() {
     public List<ProcessInstanceWithVarsDesc> queryProcessByVariablesAndTask(List<QueryParam> attributes,
                                                                             List<QueryParam> processVariables,
                                                                             List<QueryParam> taskVariables,
-                                                                            List<String> owners,
+                                                                            QueryParam owners,
```
Comment (elguardian) [inline jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/AdvanceRuntimeDataServiceImpl.java:56]: signatures of this one should not be changed as you are removing api that is already shipped within the product.
Create a new signature for this one (override)

the old one will call the new one with the QueryParam.ALL

With that you won't need to change any test. Just add a new one with ANY

Comment (ruromero) [reply jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/AdvanceRuntimeDataServiceImpl.java:65]: done

**Review by afalhambra — CHANGES_REQUESTED** [review]

**Review by afalhambra — APPROVED** [review]
Nice job @ruromero!

**Comment by afalhambra** [comment]
Jenkins retest this

File: jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/AbstractAdvanceRuntimeDataServiceImpl.java [hunk 3]
```diff
@@ -69,15 +69,23 @@ public void setEmf(EntityManagerFactory emf) {
                                                                                                            derivedTables +
                                                                                                            " WHERE  pil.processType = :processType " + globalWhere +
                                                                                                            " ORDER BY pil.processInstanceId ASC ";
-        return queryProcessUserTasksByVariables(attributes, processVariables, emptyList(), emptyList(), processType, varPrefix, queryContext, mainSQLProducer, this::collectProcessData);
+        return queryProcessUserTasksByVariables(attributes, processVariables, emptyList(), null, processType, varPrefix, queryContext, mainSQLProducer, this::collectProcessData);
```
Comment (elguardian) [inline jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/AbstractAdvanceRuntimeDataServiceImpl.java:72]: this change should not be required

File: jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/AbstractAdvanceRuntimeDataServiceImpl.java [hunk 4]
```diff
@@ -190,16 +212,28 @@ public void setEmf(EntityManagerFactory emf) {
...
+                        "           WHERE po.entity_id IN (:owners) \n" +
+                        "           GROUP BY po.task_id \n" +
+                        "           HAVING COUNT(po.entity_id) = :num_owners \n" +
+                        ") pot ON pot.task_id = task.id ");
+            } else {
+                derivedTables.append("INNER JOIN ( \n" +
+                        "           SELECT DISTINCT po.task_id \n" +
+                        "           FROM PeopleAssignments_PotOwners po \n" +
+                        "           WHERE po.entity_id IN (:owners) \n" +
+                        "           GROUP BY po.task_id \n" +
```
Comment (elguardian) [inline jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/AbstractAdvanceRuntimeDataServiceImpl.java:230]: group by is not required with the any operator.

File: jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/AdvanceRuntimeDataServiceImplTest.java [hunk 5]
```diff
@@ -352,7 +349,7 @@ public void testQueryTaskNotEqualsByVariables() {
     @Test
     public void testQueryTaskByVariablesWithOwners() {
 
-        List<String> potOwners = Collections.singletonList("katy");
+        QueryParam potOwners = new QueryParam(null, "ALL", Collections.singletonList("katy"));
```
Comment (elguardian) [inline jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/AdvanceRuntimeDataServiceImplTest.java:352]: QueryParam as others please. QueryParam.all ...QueryParam.any (Static methods)

**Comment by elguardian** [comment]
overall good... some minor changes.

File: jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/AbstractAdvanceRuntimeDataServiceImpl.java (hunk 3 above)
Comment (ruromero) [reply jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/AbstractAdvanceRuntimeDataServiceImpl.java:72]: The `owners` are now a `QueryParam`

File: jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/AbstractAdvanceRuntimeDataServiceImpl.java (hunk 4 above)
Comment (ruromero) [reply jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/AbstractAdvanceRuntimeDataServiceImpl.java:230]: done

File: jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/AdvanceRuntimeDataServiceImplTest.java (hunk 5 above)
Comment (ruromero) [reply jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/AdvanceRuntimeDataServiceImplTest.java]: Sure, that's much better, thanks. Updated tests accordingly

**Comment by ruromero** [comment]
jenkins retest this please

**Comment by elguardian** [comment]
Jenkins retest this please.

**Review by elguardian — APPROVED** [review]
waiting for CI green

**Comment by ruromero** [comment]
@elguardian I changed it so that it defaults to an `emptyList()`. It's no use in throwing an exception when it's null

**Comment by elguardian** [comment]
failures are related, have a look at them @ruromero

**Comment by ruromero** [comment]
jenkins retest this

**Comment by ruromero** [comment]
jenkins retest this please

**Comment by ruromero** [comment]
jenkins retest this once more if you don't mind, my friend

**Comment by elguardian** [comment]
jenkins retest this please

**Comment by elguardian** [comment]
jenkins retest this please

**Comment by elguardian** [comment]
jenkins retest this please

**Comment by ruromero** [comment]
jenkins retest this

```
kiegroup_jbpm/jbpm-services/jbpm-executor/src/test/java/org/jbpm/executor/impl/wih/CleanupExecutionErrorCommandWithProcessTest.java:203 | test timed out after 30000 milliseconds
```

**Comment by ruromero** [comment]
jenkins retest this
same unrelated error

**Comment by elguardian** [comment]
Jenkins retest this please

**Pull request merged**